### DISCUSSION TITLE
How to set architecture-type dhcp option?

### DISCUSSION THREAD

**Discussion by g-v-egidy** [ipxe/ipxe#271] [Q&A]
Hi,

I want to be able to use UEFI HTTP boot from within iPXE. So the DHCP server should supply a http-url instead of a filename to load via TFTP.

To support all kinds of different PXE clients, my DHCP server decides which kind of filename to supply based on architecture-type:

```
        if option architecture-type = 00:09 or option architecture-type = 00:07
        {
                filename "efi/syslinux.efi";
                option bootfile-name "efi/syslinux.efi";
        }
        else
        if option architecture-type = 00:10
        {
                filename "http://localboot.lan/httpboot/grubx64.efi";
                option bootfile-name "http://localboot.lan/httpboot/grubx64.efi";
                option vendor-class-identifier "HTTPClient";
        }
        else
        {
                filename "pxelinux.0";
                option bootfile-name "pxelinux.0";
        }
```

The architecture-type dhcp option is defined in [RFC 4578](https://tools.ietf.org/html/rfc4578) and [IANA has reserved type 0x00 0x10 for EFI HTTP boot](http://www.iana.org/assignments/dhcpv6-parameters/dhcpv6-parameters.xhtml#processor-architecture).

How can I set the architecture-type that is sent by iPXE?

The "platform" setting of iPXE seems to be something a bit different as it contains efi and pcbios, but can't differentiate between efi and efi-httpboot.

**Comment by mcb30** [comment]
iPXE will always use the URI scheme specified in `bootfile-name`.  If the value of `bootfile-name` starts with `http://`, then iPXE will perform an HTTP boot.  Unlike legacy UEFI HTTP boot, the user is not forced to configure the booting client to know in advance which URI scheme the DHCP server will select.

You can use the `user-class` option to identify iPXE clients.  A typical setup would involve handing out an iPXE binary to legacy clients, and then handing out an iPXE script to continue the rest of the boot.  For example:

```
if exists user-class and option user-class = "iPXE" {

    # iPXE client - hand out the real boot script
    filename "http://localboot.lan/httpboot/script.ipxe";

} elsif option architecture-type = 00:10 {

    # Legacy UEFI HTTP boot - hand out the UEFI iPXE binary via HTTP
    filename "http://localboot.lan/httpboot/ipxe.efi";
    option vendor-class-identifier "HTTPClient";

} elsif option architecture-type = 00:07 or option architecture-type = 00:09 {

    # Legacy UEFI PXE boot - hand out the UEFI iPXE binary via TFTP
    next-server localboot.lan;
    filename "tftpboot/ipxe.efi";

} else {

    # Legacy BIOS PXE boot - hand out the BIOS iPXE binary via TFTP
    next-server localboot.lan;
    filename "undionly.kpxe";

}
```

You can then use `script.ipxe` to control the rest of the boot process.  Within the script, you can use the `${platform}` variable to determine which platform you are running on, but this is often unnecessary.  For example, booting into a Linux kernel will work for both BIOS and UEFI using a script fragment such as

```
#!ipxe
kernel http://localboot.lan/linux/vmlinuz initrd=initrd.img
initrd http://localboot.lan/linux/initrd.img
boot
```

**Reply by g-v-egidy** [reply]
Thanks for your quick answer.

> Unlike legacy UEFI HTTP boot, the user is not forced to configure the booting client to know in advance which URI scheme the DHCP server will select.

This is something unique to iPXE. All other systems communicate their capabilities through the architecture-type option. So I guess it is quite common to differentiate the bootfile-name value based on it.

My usecase is not to use iPXE everywhere, but use it just to fix / improve some selected systems that have problems with their native PXE stack. Then I load grub (UEFI) or syslinux (classic BIOS) over the network for a boot menu or automatic provisioning. All other systems would still use their native PXE/TFTP or PXE/HTTP stack. So it would be nice if iPXE could integrate into that setup well.

> You can use the `user-class` option to identify iPXE clients.

I could identify iPXE clients this way. But not all iPXE clients will be able to handle HTTP, as you can disable this in the build config. So I can't hand out all iPXE client a http url. My dhcpd config is not just used in a small, confined network, but distributed over a whole org, customer networks and so on. So I don't know what kinds of iPXE clients are used there.

I would prefer if I could configure iPXE in way that it behaves just like traditional PXE stacks regarding the architecture-type.

Would it be a reasonable feature request to add a ```architecture-type``` setting that is of the type uint_16t? That would be communicated in the next dhcp request, autoboot and so on.

**Reply by mcb30** [reply]
> This is something unique to iPXE. All other systems communicate their capabilities through the architecture-type option. So I guess it is quite common to differentiate the bootfile-name value based on it.

In the case of legacy UEFI HTTP boot it's not communicating a capability: it's communicating the fact that the local system has been configured to **require** an HTTP URL from the DHCP server.  (Or possibly the fact that the local system just wants an IP address and is then going to ignore the DHCP filename and perform an HTTP boot using a fixed URL predefined as a UEFI boot option.  Or possibly that it's going to do something very vendor-specific, since the UEFI spec leaves this option wide open.)

> My usecase is not to use iPXE everywhere, but use it just to fix / improve some selected systems that have problems with their native PXE stack. Then I load grub (UEFI) or syslinux (classic BIOS) over the network for a boot menu or automatic provisioning.

Have you considered upgrading from grub/syslinux to use iPXE instead? :slightly_smiling_face: 

> I would prefer if I could configure iPXE in way that it behaves just like traditional PXE stacks regarding the architecture-type.

Unfortunately the architecture-type is not a well-defined value for any client that is prepared to receive an arbitrary URL from the DHCP server.

iPXE's behaviour here predates the creation of UEFI HTTP boot and is more in line with the philosophy of "just work immediately out of the box" than the somewhat broken UEFI concept of HTTP boot, so I'm definitely not changing iPXE's default behaviour.

> Would it be a reasonable feature request to add a `architecture-type` setting that is of the type uint_16t? That would be communicated in the next dhcp request, autoboot and so on.

It could be done in the same way that we currently handle the ability to override `user-class` and `vendor-class`.  I'd consider this an anti-feature, but I'd accept a pull request that was clean and had minimal code size impact.
